File: services/ai_service.py
"""
AI service for aiRekon Risk Assessment Tool
Handles all OpenAI API interactions
"""

import json
import logging
from config import get_openai_client, get_risk_conversations
from utils.prompt_builders import (
    build_overview_prompt, build_operational_prompt, build_risk_assessment_prompt,
    build_risk_conversation_system_prompt, build_event_context_message,
    build_next_risk_prompt, build_additional_risk_prompt, build_single_risk_prompt,
    build_rekon_context_prompt, build_rekon_risk_prompt, build_rekon_compliance_prompt
)
from utils.response_parsers import parse_justification_response
from services.validation_service import validate_and_format_risks, validate_and_format_single_risk

logger = logging.getLogger(__name__)

def generate_overview(event_data):
    """Generate overview paragraph for risk assessment"""
    client = get_openai_client()
    prompt = build_overview_prompt(event_data)
    
    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Generate a professional overview paragraph for the risk assessment."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=300
    )
    
    return response.choices[0].message.content.strip()

def generate_operational(event_data):
    """Generate operational considerations paragraph for risk assessment"""
    client = get_openai_client()
    prompt = build_operational_prompt(event_data)
    
    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Generate a professional operational considerations paragraph for the risk assessment."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=300
    )
    
    return response.choices[0].message.content.strip()

def generate_risks(event_data):
    """Generate risk assessment table"""
    client = get_openai_client()
    prompt = build_risk_assessment_prompt(event_data)
    
    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Generate a comprehensive risk assessment in JSON format."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=2000
    )
    
    content = response.choices[0].message.content.strip()
    
    # Parse JSON response
    try:
        risks = json.loads(content)
        return validate_and_format_risks(risks)
    except json.JSONDecodeError as e:
        logger.error(f"Failed to parse risks JSON: {e}")
        return []

def start_risk_conversation(event_data):
    """Start a new risk assessment conversation"""
    client = get_openai_client()
    risk_conversations = get_risk_conversations()
    
    conversation_id = f"event_{hash(str(event_data))}"
    
    # Initialize conversation with system prompt and event context
    system_prompt = build_risk_conversation_system_prompt()
    event_context = build_event_context_message(event_data)
    
    conversation = [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": event_context}
    ]
    
    # Store conversation
    risk_conversations[conversation_id] = {
        'messages': conversation,
        'event_data': event_data,
        'generated_risks': []
    }
    
    return conversation_id

def generate_next_risk(conversation_id, risk_number):
    """Generate the next risk in an ongoing conversation"""
    client = get_openai_client()
    risk_conversations = get_risk_conversations()
    
    if conversation_id not in risk_conversations:
        raise ValueError("Conversation not found")
    
    conversation_data = risk_conversations[conversation_id]
    previous_risks = conversation_data['generated_risks']
    
    # Build prompt for next risk
    next_risk_prompt = build_next_risk_prompt(previous_risks, risk_number)
    
    # Add to conversation
    conversation_data['messages'].append({"role": "user", "content": next_risk_prompt})
    
    # Make request to OpenAI
    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=conversation_data['messages'],
        temperature=0.7,
        max_tokens=500
    )
    
    content = response.choices[0].message.content.strip()
    
    # Add AI response to conversation
    conversation_data['messages'].append({"role": "assistant", "content": content})
    
    # Parse and validate the risk
    try:
        risk_data = json.loads(content)
        formatted_risk = validate_and_format_single_risk(risk_data, risk_number)
        
        # Store the generated risk
        conversation_data['generated_risks'].append(formatted_risk)
        
        return formatted_risk
    except json.JSONDecodeError as e:
        logger.error(f"Failed to parse risk JSON: {e}")
        return None

def generate_additional_risks(event_data, existing_risks, num_risks=3):
    """Generate additional risks for an existing assessment"""
    client = get_openai_client()

    generated_risks = []
    for i in range(num_risks):
        risk_number = len(existing_risks) + len(generated_risks) + 1

        # Build prompt for additional risk
        prompt = build_additional_risk_prompt(existing_risks + generated_risks, risk_number)

        # Make request to OpenAI
        response = client.chat.completions.create(
            model="gpt-4o-mini-2024-07-18",
            messages=[
                {
                    "role": "system",
                    "content": build_risk_conversation_system_prompt()
                },
                {
                    "role": "user",
                    "content": prompt
                }
            ],
            temperature=0.7,
            max_tokens=500
        )

        content = response.choices[0].message.content.strip()

        # Parse and validate the risk
        try:
            risk_data = json.loads(content)
            formatted_risk = validate_and_format_single_risk(risk_data, risk_number)
            generated_risks.append(formatted_risk)
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse additional risk JSON: {e}")
            continue

    return generated_risks

def generate_single_risk(event_data, risk_number, total_risks):
    """Generate a single risk for progressive loading (legacy endpoint)"""
    client = get_openai_client()
    prompt = build_single_risk_prompt(event_data, risk_number, total_risks)

    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Generate a single risk in JSON format."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=500
    )

    content = response.choices[0].message.content.strip()

    # Parse and validate the risk
    try:
        risk_data = json.loads(content)
        return validate_and_format_single_risk(risk_data, risk_number)
    except json.JSONDecodeError as e:
        logger.error(f"Failed to parse single risk JSON: {e}")
        return None

def generate_justification(event_data, field_type, field_value):
    """Generate justification for a specific field"""
    client = get_openai_client()

    # Build justification prompt based on field type
    if field_type == 'overview':
        prompt = f"""Provide a detailed justification for this overview assessment:

Event: {event_data.get('eventTitle', 'N/A')}
Overview: {field_value}

REASONING:
Explain in 2-3 concise sentences why this overview accurately captures the event's key characteristics and risk context.

SOURCES:
List 3-5 specific sources that support this assessment:"""

    elif field_type == 'operational':
        prompt = f"""Provide a detailed justification for this operational considerations assessment:

Event: {event_data.get('eventTitle', 'N/A')}
Operational Considerations: {field_value}

REASONING:
Explain in 2-3 concise sentences why these operational considerations are appropriate for this event type and context.

SOURCES:
List 3-5 specific sources that support this assessment:"""

    else:
        prompt = f"""Provide a detailed justification for this {field_type} assessment:

Event: {event_data.get('eventTitle', 'N/A')}
Assessment: {field_value}

REASONING:
Explain in 2-3 concise sentences why this assessment is appropriate and accurate.

SOURCES:
List 3-5 specific sources that support this assessment:"""

    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Provide specific justifications with credible sources."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=600
    )

    content = response.choices[0].message.content.strip()
    return parse_justification_response(content)

def generate_rekon_context(event_data, score, level):
    """Generate RekonContext Index details"""
    client = get_openai_client()
    prompt = build_rekon_context_prompt(event_data, score, level)

    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Generate exactly 3 CONCISE bullet points for the RekonContext Index. Each bullet point should be 1 short sentence (10-15 words max) about contextual complexity. Return only a JSON array of strings."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=400
    )

    content = response.choices[0].message.content.strip()

    # Parse JSON response
    try:
        details = json.loads(content)
        if not isinstance(details, list) or len(details) != 3:
            raise ValueError("Expected array of 3 strings")
        return details
    except (json.JSONDecodeError, ValueError) as e:
        logger.error(f"Failed to parse RekonContext JSON: {e}")
        # Fallback to default structure based on level
        if level == "High":
            return [
                "Large-scale event requiring extensive coordination and security planning.",
                "High public visibility with significant stakeholder and media attention.",
                "Complex regulatory requirements and multi-agency oversight needed."
            ]
        elif level == "Medium":
            return [
                "Moderate-scale event with standard planning and coordination requirements.",
                "Some public attention requiring appropriate security and safety measures.",
                "Standard regulatory compliance and local authority coordination needed."
            ]
        else:  # Low
            return [
                "Small-scale event with basic planning and minimal coordination requirements.",
                "Limited public exposure with standard safety and security considerations.",
                "Basic regulatory compliance and routine local authority notification."
            ]

def generate_rekon_risk(event_data, risks, score, level):
    """Generate RekonRisk Index details"""
    client = get_openai_client()
    prompt = build_rekon_risk_prompt(event_data, risks, score, level)

    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Generate exactly 3 CONCISE bullet points for the RekonRisk Index. Each bullet point should be 1 short sentence (10-15 words max) about risk profile. Return only a JSON array of strings."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=400
    )

    content = response.choices[0].message.content.strip()

    # Parse JSON response
    try:
        details = json.loads(content)
        if not isinstance(details, list) or len(details) != 3:
            raise ValueError("Expected array of 3 strings")
        return details
    except (json.JSONDecodeError, ValueError) as e:
        logger.error(f"Failed to parse RekonRisk JSON: {e}")
        # Fallback to default structure based on level
        if level == "High":
            return [
                "Multiple high-impact risks identified across critical operational areas.",
                "Significant likelihood of incidents requiring comprehensive mitigation strategies.",
                "Intensive monitoring and emergency response capabilities essential for safety."
            ]
        elif level == "Medium":
            return [
                "Moderate risks identified requiring standard mitigation and monitoring approaches.",
                "Some potential for incidents with manageable impact through proper planning.",
                "Regular monitoring and standard emergency procedures adequate for management."
            ]
        else:  # Low
            return [
                "Limited risks identified with low impact potential and likelihood.",
                "Minimal incident probability with standard safety measures sufficient.",
                "Basic monitoring and routine safety procedures appropriate for risk level."
            ]

def generate_rekon_compliance(event_data, risks, status):
    """Generate RekonCompliance Status details"""
    client = get_openai_client()
    prompt = build_rekon_compliance_prompt(event_data, risks, status)

    response = client.chat.completions.create(
        model="gpt-4o-mini-2024-07-18",
        messages=[
            {
                "role": "system",
                "content": "You are an expert risk assessment consultant. Generate exactly 3 CONCISE bullet points for the RekonCompliance Status. Each bullet point should be 1 short sentence (10-15 words max) about regulatory alignment (Martyn's Law, ProtectUK, ISO 27001). Return only a JSON array of strings."
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        temperature=0.7,
        max_tokens=400
    )

    content = response.choices[0].message.content.strip()

    # Parse JSON response
    try:
        details = json.loads(content)
        if not isinstance(details, list) or len(details) != 3:
            raise ValueError("Expected array of 3 strings")
        return details
    except (json.JSONDecodeError, ValueError) as e:
        logger.error(f"Failed to parse RekonCompliance JSON: {e}")
        # Fallback to default structure based on status
        if status == "Exceeds Compliance":
            return [
                "Assessment demonstrates comprehensive approach exceeding regulatory requirements.",
                "Risk identification and mitigation strategies surpass industry standards.",
                "Documentation and controls align with best practice frameworks."
            ]
        elif status == "Compliant":
            return [
                "Assessment meets essential regulatory requirements and standards.",
                "Risk management approach aligns with compliance frameworks.",
                "Basic security and safety considerations are appropriately addressed."
            ]
        else:  # Non-Compliant
            return [
                "Assessment lacks key elements required by regulatory frameworks.",
                "Critical security and safety risks are not adequately addressed.",
                "Additional risk identification and mitigation planning required."
            ]

File: risk-assessment/js/ai-service.js
/**
 * AI Service Layer for Backend API Integration
 * Handles all communication with Flask backend API
 */

class AIService {
    constructor() {
        // Use main app backend URL which has the enhanced risk generation
        this.backendURL = 'http://localhost:8085';
        this.maxRetries = 3;
        this.retryDelay = 1000; // 1 second
        this.initialized = false;
    }

    /**
     * Initialize the AI service (no API key needed - handled by backend)
     */
    initialize() {
        console.log('🤖 AI Service: Initializing backend connection...');
        this.initialized = true;
        console.log('✅ AI Service: Successfully initialized for backend communication');
    }

    /**
     * Check if AI service is properly configured
     * @returns {boolean}
     */
    isConfigured() {
        return this.initialized;
    }

    /**
     * Make a request to backend API with retry logic
     * @param {string} endpoint - API endpoint
     * @param {Object} data - Request data
     * @returns {Promise<Object>}
     */
    async makeRequest(endpoint, data) {
        if (!this.isConfigured()) {
            throw new Error('AI service not configured. Please initialize first.');
        }

        for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
            try {
                const response = await fetch(`${this.backendURL}${endpoint}`, {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify(data)
                });

                if (!response.ok) {
                    const errorData = await response.json().catch(() => ({}));
                    throw new Error(`Backend request failed: ${response.status} - ${errorData.error || response.statusText}`);
                }

                const responseData = await response.json();
                console.log(`🔧 AI Service makeRequest response for ${endpoint}:`, responseData);
                return responseData;

            } catch (error) {
                console.error(`Backend request attempt ${attempt} failed:`, error);

                if (attempt === this.maxRetries) {
                    throw error;
                }

                // Wait before retrying
                await new Promise(resolve => setTimeout(resolve, this.retryDelay * attempt));
            }
        }
    }

    /**
     * Generate contextual summary for the event using two separate prompts
     * @param {Object} eventData - Event information
     * @returns {Promise<string>}
     */
    async generateContextualSummary(eventData) {
        try {
            // Generate first paragraph
            const paragraph1 = await this.generateOverviewParagraph(eventData);

            // Generate second paragraph
            const paragraph2 = await this.generateOperationalParagraph(eventData);

            // Combine both paragraphs
            return `<p>${paragraph1}</p>\n<p>${paragraph2}</p>`;

        } catch (error) {
            console.error('Error generating contextual summary:', error);
            throw error;
        }
    }

    /**
     * Generate the first paragraph - Event overview and context
     * @param {Object} eventData - Event information
     * @returns {Promise<string>}
     */
    async generateOverviewParagraph(eventData) {
        const response = await this.makeRequest('/api/ai/generate-overview', eventData);
        return response.content.trim();
    }

    /**
     * Generate the second paragraph - Operational considerations
     * @param {Object} eventData - Event information
     * @returns {Promise<string>}
     */
    async generateOperationalParagraph(eventData) {
        const response = await this.makeRequest('/api/ai/generate-operational', eventData);
        return response.content.trim();
    }

    /**
     * Generate risk assessment table
     * @param {Object} eventData - Event information
     * @returns {Promise<Object>} - Full response with risk_data and legacy risks
     */
    async generateRiskAssessment(eventData) {
        const response = await this.makeRequest('/api/ai/generate-risks', eventData);
        console.log('🔧 AI Service received response:', response);
        return response; // Return full response instead of just response.risks
    }

    /**
     * Start a new risk assessment conversation
     * @param {Object} eventData - Event information
     * @returns {Promise<string>} - Conversation ID
     */
    async startRiskConversation(eventData) {
        const response = await this.makeRequest('/api/ai/start-risk-conversation', eventData);
        return response.conversation_id;
    }

    /**
     * Generate the next risk in an ongoing conversation
     * @param {string} conversationId - Conversation ID
     * @param {number} riskNumber - Which risk number to generate (1-based)
     * @returns {Promise<Object>}
     */
    async generateNextRisk(conversationId, riskNumber) {
        const requestData = {
            conversation_id: conversationId,
            risk_number: riskNumber
        };
        const response = await this.makeRequest('/api/ai/generate-next-risk', requestData);
        return response.risk;
    }

    /**
     * Generate a single risk for progressive loading (legacy method)
     * @param {Object} eventData - Event information
     * @param {number} riskNumber - Which risk number to generate (1-based)
     * @param {number} totalRisks - Total number of risks to generate
     * @returns {Promise<Object>}
     */
    async generateSingleRisk(eventData, riskNumber, totalRisks = 8) {
        const requestData = {
            ...eventData,
            riskNumber,
            totalRisks
        };
        const response = await this.makeRequest('/api/ai/generate-single-risk', requestData);
        return response.risk;
    }

    /**
     * Generate additional risks for an existing assessment
     * @param {string} conversationId - Conversation ID
     * @param {Array} existingRisks - Array of existing risks
     * @param {number} numAdditional - Number of additional risks to generate
     * @returns {Promise<Array>}
     */
    async generateAdditionalRisks(conversationId, existingRisks, numAdditional = 3) {
        const requestData = {
            conversation_id: conversationId,
            existing_risks: existingRisks,
            num_additional: numAdditional
        };
        const response = await this.makeRequest('/api/ai/generate-additional-risks', requestData);
        return response.risks;
    }

    /**
     * Generate justification for a specific field
     * @param {string} fieldName - Name of the field
     * @param {string} fieldValue - Value of the field
     * @param {Object} context - Additional context
     * @returns {Promise<Object>}
     */
    async generateJustification(fieldName, fieldValue, context = {}) {
        const requestData = {
            fieldName,
            fieldValue,
            context
        };

        const response = await this.makeRequest('/api/ai/generate-justification', requestData);
        return response;
    }
    /**
     * Generate RekonContext Index details
     * @param {Object} eventData - Event information
     * @param {number} score - Calculated context score (1-7)
     * @param {string} level - Context level name
     * @returns {Promise<Array>} - Array of bullet point strings
     */
    async generateRekonContextDetails(eventData, score, level) {
        const requestData = {
            ...eventData,
            score,
            level
        };
        const response = await this.makeRequest('/api/ai/generate-rekon-context', requestData);
        return response.details;
    }

    /**
     * Generate RekonRisk Index details
     * @param {Object} eventData - Event information
     * @param {Array} risks - Array of risk objects
     * @param {number} score - Calculated risk score (1-7)
     * @param {string} level - Risk level name
     * @returns {Promise<Array>} - Array of bullet point strings
     */
    async generateRekonRiskDetails(eventData, risks, score, level) {
        const requestData = {
            eventData,
            risks,
            score,
            level
        };
        const response = await this.makeRequest('/api/ai/generate-rekon-risk', requestData);
        return response.details;
    }

    /**
     * Generate RekonCompliance Status details
     * @param {Object} eventData - Event information
     * @param {Array} risks - Array of risk objects
     * @param {string} status - Compliance status
     * @returns {Promise<Array>} - Array of bullet point strings
     */
    async generateRekonComplianceDetails(eventData, risks, status) {
        const requestData = {
            eventData,
            risks,
            status
        };
        const response = await this.makeRequest('/api/ai/generate-rekon-compliance', requestData);
        return response.details;
    }






}

// Create global instance
window.aiService = new AIService();
